README.md
SAIGEgds: Scalable Implementation of Generalized mixed models in PheWAS using GDS files
====

![GPLv3](http://www.gnu.org/graphics/gplv3-88x31.png)
[GNU General Public License, GPLv3](http://www.gnu.org/copyleft/gpl.html)


## Features

Scalable implementation of generalized mixed mode with the support of Genomic Data Structure ([GDS](https://github.com/zhengxwen/SeqArray)) files and highly optimized C++ implementation. It is designed for single variant tests in large-scale phenome-wide association studies (PheWAS) with millions of variants and hundreds of thousands of samples (e.g., [UK Biobank genotype data](https://www.ukbiobank.ac.uk/scientists-3/genetic-data)), controlling for case-control imbalance and sample structure in single variant association studies.

The implementation of SAIGEgds is based on the original [SAIGE](https://github.com/weizhouUMICH/SAIGE) R package (v0.29.4.4) [Zhou et al. 2018]. It is implemented with optimized C++ codes taking advantage of sparse structure of genotypes. All of the calculation with single-precision floating-point numbers in [SAIGE](https://github.com/weizhouUMICH/SAIGE) are replaced by the double-precision calculation in SAIGEgds. SAIGEgds also implements some of the [SPAtest](https://cran.r-project.org/web/packages/SPAtest/index.html) functions in C to speed up the calculation of Saddlepoint approximation.

Benchmarks using the UK Biobank White British genotype data (N=430K) with coronary heart disease and simulated cases, show that SAIGEgds is 5 to 6 times faster than the SAIGE R package in the steps of fitting null models and p-value calculations. When used in conjunction with high-performance computing (HPC) clusters and/or cloud resources, SAIGEgds provides an efficient analysis pipeline for biobank-scale PheWAS.


## Bioconductor:

Release Version: v1.12.1 ([http://www.bioconductor.org/packages/SAIGEgds](http://www.bioconductor.org/packages/SAIGEgds))

* [Help Documents](https://rdrr.io/bioc/SAIGEgds/man)
* [Tutorial](http://www.bioconductor.org/packages/release/bioc/vignettes/SAIGEgds/inst/doc/SAIGEgds.html)
* [News](http://www.bioconductor.org/packages/release/bioc/news/SAIGEgds/NEWS)


## Package Maintainer

[Xiuwen Zheng](xiuwen.zheng@abbvie.com)


## Installation

* Requires R (≥ v3.5.0), [gdsfmt](http://www.bioconductor.org/packages/gdsfmt) (≥ v1.20.0), [SeqArray](http://www.bioconductor.org/packages/SeqArray) (≥ v1.32.0)

* Recommend [GNU GCC (≥ v6.0)](https://gcc.gnu.org), requiring C++11

* Bioconductor repository
```R
if (!requireNamespace("BiocManager", quietly=TRUE))
    install.packages("BiocManager")
BiocManager::install("SAIGEgds")
```
The `BiocManager::install()` approach may require that you build from source, i.e. `make` and compilers must be installed on your system -- see the [R FAQ](http://cran.r-project.org/faqs.html) for your operating system; you may also need to install dependencies manually.

* Development version from Github (for developers/testers only)
```R
library("devtools")
install_github("AbbVie-ComputationalGenomics/SAIGEgds")
```


## Package vignette

If the package is installed from Bioconductor repository or package rebuilding, users can start R and enter to view documentation:
```R
browseVignettes("SAIGEgds")
```


## Examples

```R
library(SeqArray)
library(SAIGEgds)

# open the GDS file for genetic relationship matrix (GRM)
grm_fn <- system.file("extdata", "grm1k_10k_snp.gds", package="SAIGEgds")
(grm_gds <- seqOpen(grm_fn))

# load phenotype
phenofn <- system.file("extdata", "pheno.txt.gz", package="SAIGEgds")
pheno <- read.table(phenofn, header=TRUE, as.is=TRUE)
head(pheno)
##   sample.id y     yy      x1 x2
## 1        s1 0 4.5542  1.5118  1
## 2        s2 0 3.7941  0.3898  1
## 3        s3 0 5.0411 -0.6212  1
## ...

# fit the null model
glmm <- seqFitNullGLMM_SPA(y ~ x1 + x2, pheno, grm_gds, trait.type="binary",
    sample.col="sample.id", num.thread=2)
## SAIGE association analysis:
## Filtering variants:
## [==================================================] 100%, completed, 0s
## Fit the null model: y ~ x1 + x2 + var(GRM)
##     # of samples: 1,000
##     # of variants: 9,976
##     using 2 threads
## ...

# close the file
seqClose(grm_gds)



################################

# open the GDS file for association testing
geno_fn <- system.file("extdata", "assoc_100snp.gds", package="SAIGEgds")
(geno_gds <- seqOpen(geno_fn))
## File: assoc_100snp.gds (10.5K)
## +    [  ] *
## |--+ description   [  ] *
## |--+ sample.id   { Str8 1000 LZMA_ra(12.6%), 625B }
## |--+ variant.id   { Int32 100 LZMA_ra(48.5%), 201B } *
## ...


# p-value calculation
assoc <- seqAssocGLMM_SPA(geno_gds, glmm, mac=10, parallel=2)
## SAIGE association analysis:
##     # of samples: 1,000
##     # of variants: 100
##     MAF threshold: NaN
##     MAC threshold: 10
##     missing threshold for variants: 0.1
##     p-value threshold for SPA adjustment: 0.05
##     variance ratio for approximation: 0.9391186
##     # of processes: 2
## [==================================================] 100%, completed, 0s
## # of variants after filtering by MAF, MAC and missing thresholds: 38
## Done.

head(assoc)
##   id chr pos rs.id ref alt AF.alt mac  num       beta        SE      pval pval.noadj converged
## 1  4   1   4   rs4   A   C 0.0100  20 1000  -0.074992  0.791685  0.924533   0.924533      TRUE
## 2 12   1  12  rs12   A   C 0.0150  30 1000  -0.091001  0.657140  0.889861   0.889861      TRUE
## 3 14   1  14  rs14   A   C 0.0375  75 1000  -0.075455  0.434152  0.862023   0.862023      TRUE
## ...


# close the file
seqClose(geno_gds)
```


## Citations

Zheng X, Davis J.Wade. SAIGEgds -- an efficient statistical tool for large-scale PheWAS with mixed models. *Bioinformatics* (2020). [DOI: 10.1093/bioinformatics/btaa731](http://dx.doi.org/10.1093/bioinformatics/btaa731).

Zhou W, Nielsen JB, Fritsche LG, Dey R, Gabrielsen ME, Wolford BN, LeFaive J, VandeHaar P, Gagliano SA, Gifford A, Bastarache LA, Wei WQ, Denny JC, Lin M, Hveem K, Kang HM, Abecasis GR, Willer CJ, Lee S. Efficiently controlling for case-control imbalance and sample relatedness in large-scale genetic association studies. *Nat Genet* (2018). Sep;50(9):1335-1341. [DOI: 10.1038/s41588-018-0184-y](https://www.nature.com/articles/s41588-018-0184-y).

Zheng X, Gogarten S, Lawrence M, Stilp A, Conomos M, Weir BS, Laurie C, Levine D. SeqArray -- A storage-efficient high-performance data format for WGS variant calls. *Bioinformatics* (2017). [DOI: 10.1093/bioinformatics/btx145](http://dx.doi.org/10.1093/bioinformatics/btx145).


## See Also

[SeqArray](https://www.bioconductor.org/packages/SeqArray): Data management of large-scale whole-genome sequence variant calls

[gds2bgen](https://github.com/zhengxwen/gds2bgen): Format conversion from bgen to gds



NEWS
CHANGES IN VERSION 1.12.5
-------------------------

    o enable RCPP_PARALLEL_USE_TBB=1 on Windows


CHANGES IN VERSION 1.12.4
-------------------------

    o fix the compiling issue on ARM64
      (see https://github.com/AbbVie-ComputationalGenomics/SAIGEgds/issues/8)


CHANGES IN VERSION 1.12.1
-------------------------

    o fix the memory issue because of using deprecated tbb::task_scheduler_init
      in RcppParallel


CHANGES IN VERSION 1.6.0
-------------------------

    o `seqFitNullGLMM_SPA()` can use imputed dosages directly without
      converting the dosages to the best-guess genotypes

    o new function `glmmHeritability()` for approximate heritability estimates


CHANGES IN VERSION 1.4.0
-------------------------

    o set-based tests: burden, ACAT-V


CHANGES IN VERSION 1.2.2
-------------------------

    o update the citation

    o work around gcc-10


CHANGES IN VERSION 1.2.0
-------------------------

    o `seqAssocGLMM_SPA()` can save to a RDS file


CHANGES IN VERSION 1.0.2
-------------------------

    o fix an issue when there is no covariate in the formula (e.g., y ~ 1)


CHANGES IN VERSION 1.0.1
-------------------------

    o add a new option for the threshold of missing variant to
      `seqAssocGLMM_SPA()`

    o change the output column 'AC.alt' to 'mac' in `seqAssocGLMM_SPA()`

    o retry model fitting at least 10 times if matrix is singular or not
      positive definite, or large variance is observed


CHANGES IN VERSION 1.0.0
-------------------------

    o first Bioconductor release


CHANGES IN VERSION 0.99.0 (10-02-2019)
-------------------------

    o package submitted to Bioconductor


CHANGES IN VERSION 0.9.10 (08-12-2019)
-------------------------

    o SAIGE algorithm implementation for quantitative outcomes


CHANGES IN VERSION 0.9.9 (07-24-2019)
-------------------------

    o add a vignette to the package

    o the default of random number generator changes in R: "Rounding" was the
      default in `RNGkind()` prior to R_3.6.0, but "Rejection" is used in
      R (>= v3.6.0). For reproduction of the results created by R (< v3.6.0),
      please use `RNGkind("Mersenne-Twister", "Inversion", "Rounding")` in
      R (>= v3.6.0)


CHANGES IN VERSION 0.9.7 (06-08-2019)
-------------------------

    o seqAssocGLMM_SPA(): load balancing in parallel


CHANGES IN VERSION 0.9.0 (01-07-2019)
-------------------------

    o first version of SAIGEgds


R/assoc_single.r
#######################################################################
#
# Package name: SAIGEgds
#
# Description:
#     Scalable and accurate implementation of generalized mixed models
# using GDS files
#
# Copyright (C) 2019-2020    Xiuwen Zheng / AbbVie-ComputationalGenomics
# License: GPL-3
#


#######################################################################
# Internal model initialization

.init_nullmod <- function(modobj, ii, maf, mac, missing, spa.pval, var.ratio,
    summac=NaN, wbeta=double(), num_wbuf=0L, acatv_mac=10)
{
    # check
    if (!is.finite(var.ratio))
        stop("Invalid variance ratio in the SAIGE model.")
    # initialize the internal model parameters
    y <- unname(modobj$obj.noK$y)
    mu <- unname(modobj$fitted.values)
    X1 <- modobj$obj.noK$X1[ii,, drop=FALSE]
    n <- length(ii)
    mobj <- list(
        maf = maf, mac = mac, missing = missing, spa.pval = spa.pval,
        tau = modobj$tau,
        y = y[ii], mu = mu[ii],
        y_mu = (y - mu)[ii],  # y - mu
        mu2 = (mu * (1 - mu))[ii],
        t_XXVX_inv = t(modobj$obj.noK$XXVX_inv[ii,, drop=FALSE]),  # K x n_samp (K << n_samp, more efficient)
        XV = modobj$obj.noK$XV[, ii, drop=FALSE],  # K x n_samp
        t_XVX_inv_XV = t(modobj$obj.noK$XXVX_inv[ii,, drop=FALSE] * modobj$obj.noK$V[ii]),  # K x n_samp
        t_X = t(X1),  # K x n_samp
        var.ratio = var.ratio,
        # buffer
        buf_dosage = double(n),
        buf_coeff = double(NROW(modobj$obj.noK$XV)),
        buf_adj_g = double(n),
        buf_index = integer(n),
        buf_B = double(n),
        buf_g_tilde = double(n),
        buf_X1 = double(NCOL(X1)),
        buf_spa = double(n+n)
    )
    # additional design matrix
    if (modobj$trait.type == "binary")
    {
        mobj$XVX <- t(X1) %*% (X1 * mobj$mu2)  # a matrix: K x K
        mobj$S_a <- colSums(X1 * mobj$y_mu)    # a vector of size K
    } else if (modobj$trait.type == "quantitative")
    {
        mobj$XVX <- t(X1) %*% X1               # a matrix: K x K
        mobj$S_a <- colSums(X1 * mobj$y_mu)    # a vector of size K
    } else
        stop("Invalid 'modobj$trait.type': ", modobj$trait.type, ".")
    # additional for aggregate tests
    mobj$summac <- summac
    mobj$acatv_mac <- acatv_mac
    mobj$buf_wbeta <- as.double(wbeta)
    mobj$buf_unitsz <- double(num_wbuf*5L)
    # output
    mobj
}

.dsnode <- function(gdsfile, nm)
{
    if (nm == "")
    {
        n <- index.gdsn(gdsfile, "genotype/data", silent=TRUE)
        if (!is.null(n))
        {
            nm <- "$dosage_alt"
        } else {
            nm <- getOption("seqarray.node_ds", "annotation/format/DS")
            n <- index.gdsn(gdsfile, nm, silent=TRUE)
            if (is.null(n))
                stop("Dosages should be stored in genotype or annotation/format/DS.")
        }
    }
    nm
}


#######################################################################
# SAIGE single variant analysis
#

seqAssocGLMM_SPA <- function(gdsfile, modobj, maf=NaN, mac=10, missing=0.1,
    dsnode="", spa.pval=0.05, var.ratio=NaN, res.savefn="", res.compress="LZMA",
    parallel=FALSE, verbose=TRUE)
{
    stopifnot(inherits(gdsfile, "SeqVarGDSClass") | is.character(gdsfile))
    stopifnot(is.numeric(maf), length(maf)==1L)
    stopifnot(is.numeric(mac), length(mac)==1L)
    stopifnot(is.numeric(missing), length(missing)==1L)
    stopifnot(is.character(dsnode), length(dsnode)==1L, !is.na(dsnode))
    stopifnot(is.character(dsnode), length(dsnode)==1L, !is.na(dsnode))
    stopifnot(is.numeric(spa.pval), length(spa.pval)==1L)
    stopifnot(is.numeric(var.ratio), length(var.ratio)==1L)
    stopifnot(is.character(res.savefn), length(res.savefn)==1L)
    if (!is.element(res.compress, c("LZMA", "LZMA_RA", "ZIP", "ZIP_RA", "none")))
        stop("`res.compress` should be one of LZMA, LZMA_RA, ZIP, ZIP_RA and none.")
    stopifnot(is.logical(verbose), length(verbose)==1L)

    if (verbose)
        cat(.crayon_inverse("SAIGE association analysis:\n"))

    # check model
    modobj <- .check_modobj(modobj, verbose)

    # GDS file
    if (is.character(gdsfile))
    {
        if (verbose)
            .cat("    open ", sQuote(gdsfile))
        gdsfile <- seqOpen(gdsfile, allow.duplicate=TRUE)
        on.exit(seqClose(gdsfile))
    } else {
        # save the filter on GDS file
        seqSetFilter(gdsfile, action="push", verbose=FALSE)
        on.exit(seqSetFilter(gdsfile, action="pop", verbose=FALSE))
    }

    # show warnings immediately
    saveopt <- options(warn=1L)
    on.exit(options(warn=saveopt$warn), add=TRUE)

    # determine the GDS node for dosages
    dsnode <- .dsnode(gdsfile, dsnode)

    # check sample ID
    seqSetFilter(gdsfile, sample.id=modobj$sample.id, verbose=FALSE)
    sid <- seqGetData(gdsfile, "sample.id")
    if (length(sid) != length(modobj$sample.id))
        stop("Some of sample IDs are not available in the GDS file.")
    ii <- match(sid, modobj$sample.id)
    if (any(is.na(ii)))
        stop("Sample IDs do not match.")

    dm <- seqSummary(gdsfile, "genotype", verbose=FALSE)$seldim
    nVariant <- dm[3L]
    if (verbose)
    {
        .cat("    # of samples: ", .pretty(dm[2L]))
        .cat("    # of variants: ", .pretty(dm[3L]))
        .cat("    MAF threshold: ", maf)
        .cat("    MAC threshold: ", mac)
        .cat("    missing threshold for variants: ", missing)
        .cat("    p-value threshold for SPA adjustment: ", spa.pval)
    }

    if (!is.finite(var.ratio))
        var.ratio <- mean(modobj$var.ratio$ratio, na.rm=TRUE)
    if (verbose)
        .cat("    variance ratio for approximation: ", var.ratio)

    if (dm[2L] <= 0) stop("No sample in the genotypic data set!")
    if (dm[3L] <= 0) stop("No variant in the genotypic data set!")

    # initialize the internal model parameters
    mobj <- .init_nullmod(modobj, ii, maf, mac, missing, spa.pval, var.ratio)

    # is forking or not?
    is_fork <- SeqArray:::.IsForking(parallel)
    njobs <- SeqArray:::.NumParallel(parallel)
    if (verbose)
        .cat("    # of processes: ", njobs)

    # initialize internally
    if (njobs<=1L || is_fork)
    {
        # forking, no need to distribute model parameters
        .Call(saige_score_test_init, mobj)
        initfun <- finalfun <- NULL
    } else {
        # pass the model parameters to each process
        if (verbose)
            cat("Distribute the model parameters to the", njobs, "processes\n")
        # initialize child processes internally
        initfun <- function(proc_id, mobj)
        {
            eval(parse(text="library(Rcpp)"))
            eval(parse(text="library(SAIGEgds)"))
            .packageEnv$modobj <- mobj
            .Call(saige_score_test_init, mobj)
        }
        # clear when exit
        finalfun <- function(proc_id, param)
        {
            .packageEnv$modobj <- NULL
            remove(modobj, envir=.packageEnv)
        }
    }

    # scan all (selected) variants
    if (modobj$trait.type == "binary")
    {
        rv <- seqParallel(parallel, gdsfile, split="by.variant",
            .initialize=initfun, .finalize=finalfun, .initparam=mobj,
            .balancing=TRUE, .bl_size=50000L, .bl_progress=verbose,
            FUN = function(f, dsnode, pverbose)
            {
                seqApply(f, dsnode, .cfunction("saige_score_test_bin"),
                    as.is="list", parallel=FALSE, .progress=pverbose,
                    .list_dup=FALSE, .useraw=NA)
            }, dsnode=dsnode, pverbose=verbose & (njobs==1L))
    } else if (modobj$trait.type == "quantitative")
    {
        rv <- seqParallel(parallel, gdsfile, split="by.variant",
            .initialize=initfun, .finalize=finalfun, .initparam=mobj,
            .balancing=TRUE, .bl_size=50000L, .bl_progress=verbose,
            FUN = function(f, dsnode, pverbose)
            {
                seqApply(f, dsnode, .cfunction("saige_score_test_quant"),
                    as.is="list", parallel=FALSE, .progress=pverbose,
                    .list_dup=FALSE, .useraw=NA)
            }, dsnode=dsnode, pverbose=verbose & (njobs==1L))
    } else
        stop("Invalid 'modobj$trait.type'.")

    # if any maf/mac filter
    if (length(rv) != nVariant)
        stop("Internal error: seqParallel() returns a vector of wrong length.")
    x <- sapply(rv, is.null)
    if (any(x))
    {
        x <- !x
        seqSetFilter(gdsfile, variant.sel=x, action="intersect", verbose=FALSE)
        rv <- rv[x]
    }
    if (verbose)
    {
        .cat(
        "# of variants after filtering by MAF, MAC and missing thresholds: ",
            .pretty(length(rv)))
    }

    # output to a GDS file?
    isfn <- !is.na(res.savefn) && res.savefn!=""
    if (isfn && grepl("\\.gds$", res.savefn, ignore.case=TRUE))
    {
        if (verbose)
            .cat("Save to ", sQuote(res.savefn), " ...")
        cm <- res.compress[1L]
        # create a GDS file
        outf <- createfn.gds(res.savefn)
        on.exit(closefn.gds(outf), add=TRUE)
        put.attr.gdsn(outf$root, "FileFormat", "SAIGE_OUTPUT")
        put.attr.gdsn(outf$root, "Version",
            paste0("SAIGEgds_", packageVersion("SAIGEgds")))
        # add function
        Add <- function(varnm, val)
            add.gdsn(outf, varnm, val, compress=cm, closezip=TRUE)
        # add sample IDs
        Add("sample.id", seqGetData(gdsfile, "sample.id"))
        # write data
        .write_gds(outf, "id", gdsfile, "variant.id", cm)
        .write_gds(outf, "chr", gdsfile, "chromosome", cm)
        .write_gds(outf, "pos", gdsfile, "position", cm)
        # rs.id
        if (!is.null(index.gdsn(gdsfile, "annotation/id", silent=TRUE)))
            .write_gds(outf, "rs.id", gdsfile, "annotation/id", cm)
        # ref and alt alleles
        Add("ref", seqGetData(gdsfile, "$ref"))
        Add("alt", seqGetData(gdsfile, "$alt"))
        # other data
        Add("AF.alt", sapply(rv, `[`, i=1L))
        Add("mac", sapply(rv, `[`, i=2L))
        Add("num", as.integer(sapply(rv, `[`, i=3L)))
        Add("beta", sapply(rv, `[`, i=4L))
        Add("SE", sapply(rv, `[`, i=5L))
        Add("pval", sapply(rv, `[`, i=6L))
        if (modobj$trait.type == "binary")
        {
            Add("p.norm", sapply(rv, `[`, i=7L))
            Add("converged", sapply(rv, `[`, i=8L)==1)
        }
        if (verbose) .cat(.crayon_inverse("Done."))
        # output
        invisible()
    } else {
        # output
        ans <- data.frame(
            id  = seqGetData(gdsfile, "variant.id"),
            chr = seqGetData(gdsfile, "chromosome"),
            pos = seqGetData(gdsfile, "position"),
            stringsAsFactors = FALSE
        )
        # add RS IDs if possible
        if (!is.null(index.gdsn(gdsfile, "annotation/id", silent=TRUE)))
            ans$rs.id <- seqGetData(gdsfile, "annotation/id")
        ans$ref <- seqGetData(gdsfile, "$ref")
        ans$alt <- seqGetData(gdsfile, "$alt")
        ans$AF.alt <- sapply(rv, `[`, i=1L)
        ans$mac <- sapply(rv, `[`, i=2L)
        ans$num  <- as.integer(sapply(rv, `[`, i=3L))
        ans$beta <- sapply(rv, `[`, i=4L)
        ans$SE   <- sapply(rv, `[`, i=5L)
        ans$pval <- sapply(rv, `[`, i=6L)
        if (modobj$trait.type == "binary")
        {
            ans$p.norm <- sapply(rv, `[`, i=7L)
            ans$converged <- as.logical(sapply(rv, `[`, i=8L))
        }

        # save file?
        if (isfn)
        {
            cm <- switch(res.compress, LZMA="xz", LZMA_RA="xz", ZIP="gzip",
                ZIP_RA="gzip", TRUE)
            if (verbose)
                .cat("Save to ", sQuote(res.savefn), " ...")
            if (grepl("\\.(rda|RData)$", res.savefn, ignore.case=TRUE))
            {
                .res <- ans
                save(.res, file=res.savefn, compress=cm)
            } else if (grepl("\\.rds$", res.savefn, ignore.case=TRUE))
            {
                saveRDS(ans, file=res.savefn, compress=cm)
            } else {
                stop("Unknown format of the output file, and it should be RData, RDS or gds.")
            }
            if (verbose) .cat(.crayon_inverse("Done."))
            invisible()
        } else {
            if (verbose) .cat(.crayon_inverse("Done."))
            ans
        }
    }
}



#######################################################################
# SAIGE single variant analysis using user-defined genotypes
#

.UserGLMM_SPA <- function(nVariant, modobj, fun, spa.pval=0.05, var.ratio=NaN,
    res.savefn="", res.compress="LZMA", parallel=FALSE, verbose=TRUE, ...)
{
    stopifnot(is.numeric(nVariant), length(nVariant)==1L, nVariant>0L)
    stopifnot(is.function(fun))
    stopifnot(is.numeric(spa.pval), length(spa.pval)==1L)
    stopifnot(is.numeric(var.ratio), length(var.ratio)==1L)
    stopifnot(is.character(res.savefn), length(res.savefn)==1L)
    if (!is.element(res.compress, c("LZMA", "LZMA_RA", "ZIP", "ZIP_RA", "none")))
        stop("`res.compress` should be one of LZMA, LZMA_RA, ZIP, ZIP_RA and none.")
    stopifnot(is.logical(verbose), length(verbose)==1L)

    # check model
    if (is.character(modobj))
    {
        stopifnot(length(modobj)==1L)
        modobj <- get(load(modobj))
    }
    stopifnot(inherits(modobj, "ClassSAIGE_NullModel"))

    # show warnings immediately
    saveopt <- options(warn=1L)
    on.exit(options(warn=saveopt$warn), add=TRUE)

    nSamp <- length(modobj$sample.id)
    if (nSamp <= 0) stop("No sample!")
    if (nVariant <= 0) stop("No variant!")

    if (verbose)
    {
        .cat(.crayon_inverse("SAIGE association analysis:"))
        .cat("    # of samples: ", .pretty(nSamp))
        .cat("    # of variants: ", .pretty(nVariant))
        .cat("    p-value threshold for SPA adjustment: ", spa.pval)
    }

    if (!is.finite(var.ratio))
        var.ratio <- mean(modobj$var.ratio$ratio, na.rm=TRUE)
    if (verbose)
        .cat("    variance ratio for approximation: ", var.ratio)

    # initialize the internal model parameters
    mobj <- .init_nullmod(modobj, ii, 0, 0, 1, spa.pval, var.ratio)

    # is forking or not?
    is_fork <- SeqArray:::.IsForking(parallel)
    njobs <- SeqArray:::.NumParallel(parallel)
    if (verbose) .cat("    # of processes: ", njobs)

    # initialize internally
    if (njobs<=1L || is_fork)
    {
        # forking, no need to distribute model parameters
        .Call(saige_score_test_init, mobj)
        initfun <- finalfun <- NULL
    } else {
        # pass the model parameters to each process
        if (verbose)
            cat("Distribute the model parameters to the", njobs, "processes\n")
        # initialize child processes internally
        initfun <- function(proc_id, mobj)
        {
            eval(.load_pkg)
            .packageEnv$modobj <- mobj
            .Call(saige_score_test_init, mobj)
        }
        # clear when exit
        finalfun <- function(proc_id, param)
        {
            .packageEnv$modobj <- NULL
            remove(modobj, envir=.packageEnv)
        }
    }

    # split
    pssplit <- SeqArray:::.file_split(nVariant, njobs, multiple=1L)

    # scan all (selected) variants
    if (modobj$trait.type == "binary")
    {
        rv <- seqParallel(parallel, NULL, split="none",
            .initialize=initfun, .finalize=finalfun, .initparam=mobj,
            FUN = function(fun, pssplit, verbose)
            {
                i <- SeqArray:::process_index
                st <- pssplit[[1L]][i] - 1L
                n <- pssplit[[2L]][i]
                ans <- vector("list", n)
                f <- .cfunction("saige_score_test_bin")
                if (n > 0L)
                {
                    pg <- if (verbose && i==1L) SeqArray:::.seqProgress(n, 1L) else NULL
                    for (j in seq_len(n))
                    {
                        v <- f(fun(j+st))
                        if (!is.null(v)) ans[[j]] <- v
                        SeqArray:::.seqProgForward(pg, 1L)
                    }
                    remove(pg)
                }
                ans
            }, fun=fun, pssplit=pssplit, verbose=verbose)
    } else if (modobj$trait.type == "quantitative")
    {
        rv <- seqParallel(parallel, NULL, split="none",
            .initialize=initfun, .finalize=finalfun, .initparam=mobj,
            FUN = function(fun, pssplit, verbose)
            {
                i <- SeqArray:::process_index
                st <- pssplit[[1L]][i] - 1L
                n <- pssplit[[2L]][i]
                ans <- vector("list", n)
                f <- .cfunction("saige_score_test_quant")
                if (n > 0L)
                {
                    pg <- if (verbose && i==1L) SeqArray:::.seqProgress(n, 1L) else NULL
                    for (j in seq_len(n))
                    {
                        v <- f(fun(j+st))
                        if (!is.null(v)) ans[[j]] <- v
                        SeqArray:::.seqProgForward(pg, 1L)
                    }
                    remove(pg)
                }
                ans
            }, fun=fun, pssplit=pssplit, verbose=verbose)
    } else {
        stop("Invalid 'modobj$trait.type'.")
    }

    # if any maf/mac filter
    if (length(rv) != nVariant)
        stop("Internal error: seqParallel() returns a vector of wrong length.")
    x <- sapply(rv, is.null)
    ii <- which(!x)
    if (any(x)) rv <- rv[ii]
    if (verbose)
    {
        cat("# of variants after filtering by MAF, MAC and missing thresholds: ",
            .pretty(length(rv)), "\n", sep="")
    }

    # output to a GDS file?
    isfn <- !is.na(res.savefn) && res.savefn!=""
    if (isfn && grepl("\\.gds$", res.savefn, ignore.case=TRUE))
    {
        if (verbose)
            .cat("Save to ", sQuote(res.savefn), " ...")
        cm <- res.compress[1L]
        # create a GDS file
        outf <- createfn.gds(res.savefn)
        on.exit(closefn.gds(outf), add=TRUE)
        put.attr.gdsn(outf$root, "FileFormat", "SAIGE_OUTPUT")
        put.attr.gdsn(outf$root, "Version",
            paste0("SAIGEgds_", packageVersion("SAIGEgds")))
        # add function
        Add <- function(varnm, val)
            add.gdsn(outf, varnm, val, compress=cm, closezip=TRUE)
        # write data
        Add("sample.id", modobj$sample.id)
        Add("id", ii)
        Add("AF.alt", sapply(rv, `[`, i=1L))
        Add("mac", sapply(rv, `[`, i=2L))
        Add("num", as.integer(sapply(rv, `[`, i=3L)))
        Add("beta", sapply(rv, `[`, i=4L))
        Add("SE", sapply(rv, `[`, i=5L))
        Add("pval", sapply(rv, `[`, i=6L), compress=cm, closezip=TRUE)
        if (modobj$trait.type == "binary")
        {
            Add("p.norm", sapply(rv, `[`, i=7L))
            Add("converged", sapply(rv, `[`, i=8L)==1L)
        }
        if (verbose) cat(.crayon_inverse("Done.\n"))
        # output
        invisible()
    } else {
        # output
        ans <- data.frame(
            id = ii,
            AF.alt = sapply(rv, `[`, i=1L),
            mac = sapply(rv, `[`, i=2L),
            num = as.integer(sapply(rv, `[`, i=3L)),
            beta = sapply(rv, `[`, i=4L),
            SE   = sapply(rv, `[`, i=5L),
            pval = sapply(rv, `[`, i=6L))
        if (modobj$trait.type == "binary")
        {
            ans$p.norm <- sapply(rv, `[`, i=7L)
            ans$converged <- as.logical(sapply(rv, `[`, i=8L))
        }

        # save file?
        if (isfn)
        {
            if (grepl("\\.(rda|RData)$", res.savefn, ignore.case=TRUE))
            {
                if (verbose)
                    .cat("Save to ", sQuote(res.savefn), " ...")
                cm <- switch(res.compress, LZMA="xz", LZMA_RA="xz",
                    ZIP="gzip", ZIP_RA="gzip", none="gzip", TRUE)
                .res <- ans
                save(.res, file=res.savefn, compress=cm)
                if (verbose) cat(.crayon_inverse("Done.\n"))
                invisible()
            } else {
                stop("Unknown format of the output file, and it should be RData or gds.")
            }
        } else {
            if (verbose) cat(.crayon_inverse("Done.\n"))
            ans
        }
    }
}
